batchflow/models/tf/layers/conv_block.py

```diff
@@ -159,33 +143,296 @@ def conv_block(inputs, layout='', filters=0, kernel_size=3, name=None,
 
     ::
 
-        x = conv_block(x, 'ca ca ca nd', [32, 32, 64], [5, 3, 3], strides=[1, 1, 2], dropout_rate=.15)
+        x = ConvBlock('ca ca ca nd', [32, 32, 64], [5, 3, 3], strides=[1, 1, 2], dropout_rate=.15)(x)
 
     A residual block::
 
-        x = conv_block(x, 'R nac nac +', [16, 16, 64], [1, 3, 1])
+        x = ConvBlock('R nac nac +', [16, 16, 64], [1, 3, 1])(x)
 
     """
-    tensor = _conv_block(inputs, layout, filters, kernel_size, name,
-                         strides, padding, data_format, dilation_rate, depth_multiplier,
-                         activation, pool_size, pool_strides, dropout_rate, is_training,
-                         **kwargs)
-    return tensor
-
-
-def upsample(inputs, factor=None, shape=None, layout='b', name='upsample', **kwargs):
-    """ Upsample inputs with a given factor
+    C_LAYERS = {
+        'a': 'activation',
+        'R': 'residual_start',
+        '+': 'residual_end',
+        '.': 'residual_end',
+        'f': 'dense',
+        'c': 'conv',
+        't': 'transposed_conv',
+        'C': 'separable_conv',
+        'T': 'separable_conv_transpose',
+        'p': 'pooling',
+        'v': 'pooling',
+        'P': 'global_pooling',
+        'V': 'global_pooling',
+        'n': 'batch_norm',
+        'd': 'dropout',
+        'D': 'alpha_dropout',
+        'O': 'dropblock',
+        'm': 'mip',
+        'A': 'residual_bilinear_additive',
+        'b': 'resize_bilinear',
+        'B': 'resize_bilinear_additive',
+        'N': 'resize_nn',
+        'X': 'subpixel_conv'
+    }
+
+    FUNC_LAYERS = {
+        'activation': None,
+        'residual_start': None,
+        'residual_end': None,
+        'dense': K.Dense,
+        'conv': Conv,
+        'transposed_conv': ConvTranspose,
+        'separable_conv': SeparableConv,
+        'separable_conv_transpose': SeparableConvTranspose,
+        'pooling': Pooling,
+        'global_pooling': GlobalPooling,
+        'batch_norm': K.BatchNormalization,
+        'dropout': K.Dropout,
+        'alpha_dropout': K.AlphaDropout,
+        'dropblock': Dropblock,
+        'mip': Mip,
+        'residual_bilinear_additive': None,
+        'resize_bilinear': ResizeBilinear,
+        'resize_bilinear_additive': ResizeBilinearAdditive,
+        'resize_nn': ResizeNn,
+        'subpixel_conv': SubpixelConv
+    }
+
+    LAYER_KEYS = ''.join(list(C_LAYERS.keys()))
+    GROUP_KEYS = (
+        LAYER_KEYS
+        .replace('t', 'c')
+        .replace('C', 'c')
+        .replace('T', 'c')
+        .replace('v', 'p')
+        .replace('V', 'P')
+        .replace('D', 'd')
+        .replace('O', 'd')
+        .replace('A', 'b')
+        .replace('B', 'b')
+        .replace('N', 'b')
+        .replace('X', 'b')
+    )
+
+    C_GROUPS = dict(zip(LAYER_KEYS, GROUP_KEYS))
```

Comment:
```suggestion:-16+14
    C_GROUPS = {k: k for k in C_LAYERS}
    C_GROUPS.update({
        't': 'c',
        'C': 'c',
        'T': 'c',
        'v': 'p',
        'V': 'P',
        'D': 'd',
        'O': 'd',
        'A': 'b',
        'B': 'b',
        'N': 'b',
        'X': 'b',
    })
```

Reply:
Is that necessary?

Reply:
No, it's up to you. IMO my proposal is just more clear (and 2 lines shorter).